File: tests/verify.py
#!/usr/bin/env python3
"""前后端桥接一致性验证（20 项）"""
import re, sys

def read(p):
    with open(p, encoding='utf-8') as f: return f.read()

java = read('android/app/src/main/java/com/ncmconverter/app/WebAppInterface.java')
main = read('android/app/src/main/java/com/ncmconverter/app/MainActivity.java')
js = read('www/app.js')

checks = []
def check(name, cond): checks.append((name, bool(cond)))

# ---- Java 端：注册名 + 方法签名 ----
check("Java: 注册名为 'Android'", 'addJavascriptInterface(' in main and '"Android"' in main)
check("Java: pickFiles() 无参", 'public void pickFiles()' in java)
check("Java: pickFolder() 无参", 'public void pickFolder()' in java)
check("Java: saveAudio 4 参数 (String,String,String,String)",
      re.search(r'public void saveAudio\(String[^)]*String[^)]*String[^)]*String', java) is not None)
save_params = re.search(r'saveAudio\((String[^)]*)\)', java)
check("Java: saveAudio 参数顺序 folderUri,baseName,ext,base64Data",
      'folderUri' in (save_params.group(1) if save_params else '') and 'base64Data' in (save_params.group(1) if save_params else ''))

# ---- Java 端：回调 JS 的方法名 ----
check("Java: 回调 window.onFilesPicked", 'window.onFilesPicked' in main)
check("Java: 回调 window.onFolderPicked", 'window.onFolderPicked' in main)
check("Java: 传 base64 字段", '"base64"' in main)
check("Java: 传 uri + name 字段", '"uri"' in main and '"name"' in main)

# ---- JS 端：调用方式 ----
check("JS: window.Android 对象", 'window.Android' in js or 'Android = window.Android' in js)
check("JS: Android.pickFiles()", 'Android.pickFiles()' in js)
check("JS: Android.pickFolder()", 'Android.pickFolder()' in js)
check("JS: Android.saveAudio 4 参数", js.count('Android.saveAudio(') > 0 and
      all(len([a for a in c.split('(')[1].split(')')[0].split(',') if a.strip()]) == 4
          for c in re.findall(r'Android\.saveAudio\([^)]*\)', js)))
save_calls = re.findall(r'Android\.saveAudio\(([^)]*)\)', js)
check("JS: saveAudio 参数顺序 (folderUri, baseName, ext, b64)",
      any('folderUri' in c or 'folder' in c.lower() for c in save_calls))

# ---- JS 端：回调定义 ----
check("JS: window.onFilesPicked 定义", 'window.onFilesPicked' in js)
check("JS: window.onFolderPicked 定义", 'window.onFolderPicked' in js)
check("JS: onFilesPicked 用 JSON.parse", 'JSON.parse' in js)
check("JS: 读取 f.base64", 'base64' in js)
check("JS: 读取 folder.uri + folder.name", 'folder.uri' in js and 'folder.name' in js)

# ---- 应用标识 ----
manifest = read('android/app/src/main/AndroidManifest.xml')
gradle = read('android/app/build.gradle')
check("应用名 NCM 解密器 Pro", 'NCM 解密器 Pro' in manifest)
check("包名 com.ncmdecoder.pro", 'com.ncmdecoder.pro' in gradle)
check("版本 2.0", 'versionName "2.0"' in gradle)

# ---- Workflow ----
wf = read('.github/workflows/build-apk.yml')
check("Workflow: artifact 名带 -pro", 'NCM-Decoder-Pro' in wf)
check("Workflow: tag v2.0-pro", 'v2.0-pro' in wf)

passed = sum(1 for _, c in checks if c)
print("=" * 60)
for name, c in checks:
    print(("  ✅ " if c else "  ❌ ") + name)
print("=" * 60)
print("结果: %d/%d" % (passed, len(checks)))
sys.exit(0 if passed == len(checks) else 1)

File: tests/verify.js
/**
 * 验证：前端 app.js 的 Android 调用 与 Java WebAppInterface 方法签名是否对齐
 * 用法: node verify.js
 */
const fs = require('fs');

const appJs = fs.readFileSync('www/app.js', 'utf-8');
const javaSrc = fs.readFileSync('android/app/src/main/java/com/ncmconverter/app/WebAppInterface.java', 'utf-8');
const mainJava = fs.readFileSync('android/app/src/main/java/com/ncmconverter/app/MainActivity.java', 'utf-8');

let pass = 0, fail = 0;
function check(cond, msg) { if (cond) { pass++; console.log('  ✅ ' + msg); }
                           else { fail++; console.log('  ❌ ' + msg); } }

console.log('=== 接口对齐验证 ===\n');

// 1. JS 端注册名 ===
console.log('1. JS 桥接注册名');
check(/window\.Android\s*=/m.test(appJs) || /Android\s*=\s*window\.Android/m.test(appJs) || /const Android = window\.Android/.test(appJs),
  '前端使用 window.Android');

// 2. Java 端注册名
console.log('2. Java 桥接注册名');
check(/addJavascriptInterface\([^,]+,\s*"Android"\)/.test(mainJava),
  'Java 注册名为 "Android"');

// 3. 方法签名对齐
console.log('3. 方法签名对齐');
const jsCalls = {
  pickFiles: /Android\.pickFiles\(\)/.test(appJs),
  pickFolder: /Android\.pickFolder\(\)/.test(appJs),
  saveAudio: /Android\.saveAudio\(/.test(appJs)
};
check(jsCalls.pickFiles, 'JS 调用 Android.pickFiles()');
check(jsCalls.pickFolder, 'JS 调用 Android.pickFolder()');
check(jsCalls.saveAudio, 'JS 调用 Android.saveAudio(...)');

// Java 端方法
check(/public void pickFiles\(\)/.test(javaSrc), 'Java 定义 pickFiles()');
check(/public void pickFolder\(\)/.test(javaSrc), 'Java 定义 pickFolder()');

// saveAudio 参数：Java 必须是 (String, String, String, String)
const saveAudioMatch = javaSrc.match(/public void saveAudio\(([^)]+)\)/);
if (saveAudioMatch) {
  const params = saveAudioMatch[1].trim();
  check(params === 'String folderUri, String baseName, String ext, String base64Data',
    'saveAudio 签名: (String, String, String, String) — 实际: ' + params);
} else {
  fail++; console.log('  ❌ Java 未定义 saveAudio');
}

// 4. JS 端 saveAudio 调用参数顺序（用正则按引号外逗号分割）
console.log('4. saveAudio 调用参数顺序');
const saveCall = appJs.match(/Android\.saveAudio\(([^)]+)\)/);
if (saveCall) {
  // 简单的括号匹配分割（忽略字符串内的逗号）
  let depth = 0, start = 0;
  const args = [];
  const raw = saveCall[1];
  for (let i = 0; i < raw.length; i++) {
    const c = raw[i];
    if (c === '(') depth++;
    else if (c === ')') depth--;
    else if (c === "'" || c === '"') {
      // 跳过字符串
      const q = c;
      i++;
      while (i < raw.length && raw[i] !== q) i++;
    } else if (c === ',' && depth === 0) {
      args.push(raw.slice(start, i).trim());
      start = i + 1;
    }
  }
  args.push(raw.slice(start).trim());
  check(args.length === 4, `参数个数 = 4 (实际 ${args.length})`);
  check(args[0].includes('folder') || args[0].includes('uri'), `第1参=uri (${args[0]})`);
  check(args[1].includes('baseName') || args[1].includes('Name'), `第2参=baseName (${args[1]})`);
  check(args[2] === 'ext', `第3参=ext (${args[2]})`);
  check(args[3].includes('b64') || args[3].includes('base64'), `第4参=base64`);
}

// 5. JS→Java 回调
console.log('5. Java→JS 回调');
check(/window\.setFiles/.test(mainJava) && /setFiles/.test(appJs),
  'setFiles 回调对齐');
check(/window\.setFolder/.test(mainJava) && /setFolder/.test(appJs),
  'setFolder 回调对齐');

// 6. 文件选择回调：onShowFileChooser
console.log('6. WebChromeClient 重写');
check(/onShowFileChooser/.test(mainJava), '重写 onShowFileChooser');

// 7. 权限
console.log('7. 存储权限');
check(/READ_EXTERNAL_STORAGE/.test(mainJava) || /READ_EXTERNAL_STORAGE/.test(
  fs.readFileSync('android/app/src/main/AndroidManifest.xml', 'utf-8')), '声明读存储权限');

console.log(`\n=== 结果: ${pass} 通过, ${fail} 失败 ===`);
process.exit(fail > 0 ? 1 : 0);
